test/topology/conftest.py

```diff
@@ -216,10 +216,18 @@ def fails_without_consistent_cluster_management(request, check_pre_consistent_cl
 
 # "random_tables" fixture: Creates and returns a temporary RandomTables object
 # used in tests to make schema changes. Tables are dropped after test finishes
-# unless the cluster is dirty.
+# unless the cluster is dirty or the test has failed.
 @pytest.fixture(scope="function")
 async def random_tables(request, manager):
     tables = RandomTables(request.node.name, manager, unique_name())
     yield tables
-    if not await manager.is_dirty():
+
+    # Don't drop tables at the end if we failed or the cluster is dirty - it may be impossible
+    # (e.g. the cluster is completely dead) and it doesn't matter (we won't reuse the cluster
+    # anyway).
+    # The cluster will be marked as dirty if the test failed, but that happens
+    # at the end of `manager` fixture which we depend on (so these steps will be
+    # executed after us) - so at this point, we need to check for failure ourselves too.
+    failed = request.node.stash[FAILED_KEY]
```

Comment:
I'm not enthusiastic about this "stash" thing, that adds yet another complication.
Why can't this code simply trust manager.is_dirty()?
I.e., what I mean is that if you want a failed test to mark the cluster dirty, just do exactly that - when a test fails. mark the cluster dirty. Don't add yet another FAILED_KEY flag.

Or am I missing something?

Reply:
@kbr-scylla ^^

Reply:
> when a test fails, mark the cluster dirty.

And the stash thing is used to implement exactly that - to mark the cluster as dirty when the test fails.

You might suggest to do it in test.py, after run_test - however, this is executed after every test case in the file is executed. The `pytest_runtest_makereport` hook allows us to inspect the result of a single **test case**.

Reply:
I didn't say you shouldn't use pytest_runtest_makereport (in cql-pytest, I used a autouse fixture for the same purpose, IIUC), but didn't understand why that "stash" thing is needed - why can't makereport (or some fixture) mark the cluster dirty immediately instead of leaving some flag in some stash so that other code later marks the cluster dirty?

Anyway, I just **thought** this could make the code easier to understand. I didn't actually try to code this so maybe I'm missing something...

Reply:
> why can't makereport (or some fixture) mark the cluster dirty immediately instead of leaving some flag in some stash so that other code later marks the cluster dirty?

I think that:
- the hook does not have access
- the `manager` fixture has access to the cluster, but it doesn't know whether the test failed

so we combine the two - using a stash which is stored inside pytest's `request.node` - and pass information about the failed test from the hook to the `manager` fixture.

This is the approach suggested by pytest docs
https://docs.pytest.org/en/latest/example/simple.html#making-test-result-information-available-in-fixtures

Reply:
> the hook does not have access

... to the cluster

Reply:
>     * the hook does not have access to the cluster

Interesting point. You could make the cluster "global" (a fixture that the hook or autouse fixture uses, just like the test code). This is more-or-less what cql-pytest does (it has a "cql" fixture that everyone uses).
I agree that maybe it's nicer not to have any "global" fixtures.

>     * the `manager` fixture has access to the cluster, but it doesn't know whether the test failed

I suggested that maybe it doesn't matter if the test failed - you just need to test the cluster's liveness at the end of each test, and mark it dirty if it's not live.

But I think that as the Perl moto said, "there is more than one way to do it" :-) Let's accept your way for now, and go back to debating the different approaches if there'll be problems with this approach.